lib/deep_nest.rb

```diff
@@ -2,7 +2,121 @@
 
 require_relative "deep_nest/version"
 
+##
+# Primary namespace for the deep_nest gem.
 module DeepNest
-  class Error < StandardError; end
-  # Your code goes here...
+  class << self
+    ##
+    # Returns a deep copy of the passed object
+    #
+    # @param obj [Object] The object to be deep copied.
+    #
+    # @return [obj] The deep copy of the passed object.
+    def deep_dup(obj)
+      case obj
+      when Array
+        obj.map { |x| deep_dup(x) }
+      when Hash
+        obj.transform_values { |v| deep_dup(v) }
+      else
+        obj.dup
+      end
+    end
+
+    ##
+    # Returns a hash with the passed hashes recursively merged. An optional block can be passed to merge values.
+    #
+    # @param hash1 [Hash] The first hash to be merged.
+    #
+    # @param hash2 [Hash] The second hash to be merged.
+    #
+    # @yield [&block] Operation to merge values of the hashes.
+    #
+    # @return [Hash] The hash with recursively merged hashes.
+    def deep_merge(hash1, hash2, &block)
+      hash1.merge(hash2) do |k, v1, v2|
+        if v1.is_a?(Hash) && v2.is_a?(Hash)
+          deep_merge(v1, v2, &block)
+        elsif block_given?
+          block.call(k, v1, v2)
+        else
+          v2
+        end
+      end
+    end
+
+    ##
+    # Returns true if the passed objects are the same object, false otherwise.
+    #
+    # @param obj1 [Object] First object to be compared.
+    #
+    # @param obj2 [Object] Second object to be compared.
+    #
+    # @return [true] If passed objects are the same object.
+    #
+    # @return [false] If passed objects are not the same object.
+    def deep_equal?(obj1, obj2)
+      obj1.object_id == obj2.object_id
+    end
```

Comment:
This is doing the opposite of what we want - it's doing a more strict and shallow comparison than Ruby would do by default. What we really want is to be able to pass two nested structures, and know that they are equal in structure and values.

This will likely be the most challenging method of the bunch. You'll need to check equality in both directions.  You'll have to check that both objects are the same type. And for hashes you'll have to ensure that they have the same list of keys - that neither hash has keys that the other does not.